### PR TITLE
caching featureflags

### PR DESCRIPTION
**Pull request by brettsam**
### Issue describing the changes in this PR

Caching `FeatureFlags` values so we do not read and parse environment variables every time we access. This is a "papercut" perf improvement.

Benchmark details here -- the "Placeholder - True" runs are what we do today, with no caching. 

```
|                Method | Placeholder | FeatureFlagsValue |       Mean |     Error |    StdDev |     Median |        Min |        Max |  Gen 0 | Allocated |
|---------------------- |------------ |------------------ |-----------:|----------:|----------:|-----------:|-----------:|-----------:|-------:|----------:|
| FeatureFlag_IsEnabled |       False |                 ? |   9.009 ns | 0.3395 ns | 0.3486 ns |   9.002 ns |   8.454 ns |   9.874 ns |      - |         - |
| FeatureFlag_IsEnabled |       False |              Test |  14.097 ns | 0.0434 ns | 0.0385 ns |  14.104 ns |  14.029 ns |  14.159 ns |      - |         - |
| FeatureFlag_IsEnabled |       False |  Test,Test2,Test3 |  15.082 ns | 0.1724 ns | 0.1528 ns |  15.005 ns |  14.930 ns |  15.365 ns |      - |         - |
| FeatureFlag_IsEnabled |        True |                 ? | 248.394 ns | 4.2557 ns | 3.7726 ns | 246.317 ns | 244.996 ns | 255.852 ns | 0.0081 |      88 B |
| FeatureFlag_IsEnabled |        True |              Test | 266.581 ns | 0.7634 ns | 0.5960 ns | 266.457 ns | 265.624 ns | 267.514 ns | 0.0118 |     120 B |
| FeatureFlag_IsEnabled |        True |  Test,Test2,Test3 | 313.294 ns | 1.6296 ns | 1.4446 ns | 312.875 ns | 311.538 ns | 316.273 ns | 0.0251 |     256 B |
```

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

**Review by mathewc — APPROVED** [review]

File: src/WebJobs.Script/Config/FeatureFlags.cs [hunk 1]
```diff
@@ -12,18 +14,52 @@ namespace Microsoft.Azure.WebJobs.Script.Config
     /// </summary>
     public static class FeatureFlags
     {
+        private static readonly object _cacheLock = new();
```
Comment (jviau) [inline src/WebJobs.Script/Config/FeatureFlags.cs:17]: For private static, do we use `s_` prefix? Should we use `s_` to differentiate from instance?

File: src/WebJobs.Script/Config/FeatureFlags.cs [hunk 2]
```diff
@@ -12,18 +14,52 @@ namespace Microsoft.Azure.WebJobs.Script.Config
...
+        }
+
+        private static string[] GetFeatureFlags(IEnvironment environment)
+        {
+            string featureFlags = environment.GetEnvironmentVariable(EnvironmentSettingNames.AzureWebJobsFeatureFlags) ?? string.Empty;
+            return featureFlags.Split(',');
         }
     }
-}
+}
```
Comment (jviau) [inline src/WebJobs.Script/Config/FeatureFlags.cs:65]: nit: add empy newline at end of file

**Review by jviau — APPROVED** [review]

**Comment by brettsam** [comment]
Ok -- pause on this -- this static caching makes tests basically un-usable. Every class that accesses feature flags now needs to clear the cache after any test.  And b/c it's static, it's not even clear what is using this internally. Lots of failures.

I'm refactoring to make this an actual DI service so that it's a lot more testable and usable. There is at least one spot where this won't be possible without major changes so I'm going to let that spot retain its static access (and that spot doesn't need caching -- it's not a hot path).

**Comment by jviau** [comment]
@brettsam if you are looking to make it DI, take a look at https://github.com/microsoft/FeatureManagement-Dotnet.

**Review by liliankasem — APPROVED** [review]

**Comment by brettsam** [comment]
This isn't going to work the way that I implemented it... I was hoping it would be a quick win, but it became larger than anticipated. We'll need to come back to this next version.

**Pull request closed without merging**